executor/common_linux.h

```diff
@@ -522,6 +538,61 @@ static void netlink_devlink_netns_move(const char* bus_name, const char* dev_nam
 	close(sock);
 }
 
+static struct nlmsg nlmsg2;
+
+static void initialize_devlink_ports(const char* bus_name, const char* dev_name,
+				     const char* netdev_prefix)
+{
+	struct genlmsghdr genlhdr;
+	int len, total_len, id, err, offset;
+	uint16 netdev_index;
+
+	int sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_GENERIC);
+	if (sock == -1)
+		fail("socket(AF_NETLINK) failed\n");
+
+	int rtsock = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+	if (rtsock == -1)
+		fail("socket(AF_NETLINK) failed");
+
+	id = netlink_devlink_id_get(&nlmsg, sock);
+	if (id == -1)
+		goto error;
+
+	memset(&genlhdr, 0, sizeof(genlhdr));
+	genlhdr.cmd = DEVLINK_CMD_PORT_GET;
+	netlink_init(&nlmsg, id, NLM_F_DUMP, &genlhdr, sizeof(genlhdr));
+	netlink_attr(&nlmsg, DEVLINK_ATTR_BUS_NAME, bus_name, strlen(bus_name) + 1);
+	netlink_attr(&nlmsg, DEVLINK_ATTR_DEV_NAME, dev_name, strlen(dev_name) + 1);
+
+	err = netlink_send_ext(&nlmsg, sock, id, &total_len);
+	if (err) {
+		debug("netlink: failed to get port get reply: %s\n", strerror(err));
+		goto error;
+	}
+
+	offset = 0;
+	netdev_index = 0;
+	while ((len = netlink_next_msg(&nlmsg, offset, total_len)) != -1) {
+		struct nlattr* attr = (struct nlattr*)(nlmsg.buf + offset + NLMSG_HDRLEN + NLMSG_ALIGN(sizeof(genlhdr)));
+		for (; (char*)attr < nlmsg.buf + offset + len; attr = (struct nlattr*)((char*)attr + NLMSG_ALIGN(attr->nla_len))) {
+			if (attr->nla_type == DEVLINK_ATTR_NETDEV_NAME) {
+				char* port_name;
+				char netdev_name[IFNAMSIZ];
+				port_name = (char*)(attr + 1);
+				snprintf(netdev_name, sizeof(netdev_name), "%s%d", netdev_prefix, netdev_index);
+				netlink_device_change(&nlmsg2, rtsock, port_name, true, 0, 0, 0, netdev_name);
+				break;
+			}
+		}
+		offset += len;
+		netdev_index++;
```

Comment:
How many do we expect to have? In the next commit we only use "netdevsim0" and "netpci0" in devnames. So is it just 1?

Reply:
netdevsim spawns 4 ports - Netdevices are renamed to: netdevsim0-netdevsim3
For netpci, depends on the HW. It is 32 in my testing environment.

I put only "netdevsim0" and "netpci0" to socket.txt as it is fixed array. Do you suggest to do that dynamically? I think it is fine to work with 1 netdev in multi-netdev devices.

Reply:
You know better. If using only "netdevsim0" is good enough, let's do that.
But then strictly saying, we could rename just 1 device, right? And then we also don't need the nested netlink message functionality (writing netlink messages while reading previous netlink message).

Reply:
Hmm, I think that it would be odd to rename one and let the others as they are. The "nesting' is not such a big deal. At least it is clear what nlmsg struct the function is working with avoiding the direct access to global static struct. Eventually, I would like to construct the array dynamically, also because there are likely going to be other netdevices, like Eswitch port representors for SR-IOV etc.

Reply:
OK, let's go with nesting.
But please keep in mind that for this code we should not necessary do the same trade offs as we would normally do for typical production code. The main concern is that this code is inlined into every reproducer and a developer needs to read it for lots of reported bugs, so dumping a full-fledged framework with all the proper structuring, etc may not be the right choice. The second is related to performance -- this setup code is executed much more frequently than device setup in normal contexts (usually it's one off, so nobody usually really cares about setup cost).